Source/Core/Core/HW/GCMemcardDirectory.cpp

```diff
@@ -186,10 +188,47 @@ GCMemcardDirectory::GCMemcardDirectory(const std::string& directory, int slot, u
 	m_dir1.fixChecksums();
 	m_dir2 = m_dir1;
 	m_bat2 = m_bat1;
+
+	m_flush_thread = std::thread(&GCMemcardDirectory::FlushThread, this);
+}
+
+void GCMemcardDirectory::FlushThread()
+{
+	if (!SConfig::GetInstance().m_LocalCoreStartupParameter.bEnableMemcardSaving)
+	{
+		return;
+	}
+
+
+	Common::SetCurrentThreadName(
+		StringFromFormat("Memcard%x-Flush", card_index).c_str());
+
+	while (true)
+	{
+		// no-op until signalled
+		m_sleep_trigger.Wait();
+
+		if (m_exiting)
+		{
+			m_exiting = false;
+			return;
+		}
+		// no-op as long as signalled within flush_interval
+		while (m_flush_trigger.WaitFor(flush_interval)){}
```

Comment:
Worst case scenario: ~GCMemcardDirectory() is called by another thread while you just started waiting in m_flush_trigger. This would cause a freeze of the destructing thread until the flush_interval is elapsed (ignoring the m_exiting issue). I would recommend adding a check for m_exiting inside the wait loop here, and wake/set m_flush_trigger in ~GCMemcardDirectory() together with m_sleep_trigger, to make sure this never happens.